pcapgo/ngread.go

```diff
@@ -0,0 +1,599 @@
+}
+
+// DefaultNgReaderOptions provides sane defaults for a pcapng reader.
+var DefaultNgReaderOptions = NgReaderOptions{}
+
+// NgReader wraps an underlying bufio.NgReader to read packet data in pcapng.
+type NgReader struct {
+	r                 *bufio.Reader
+	options           NgReaderOptions
+	sectionInfo       NgSectionInfo
+	linkType          layers.LinkType
+	ifaces            []NgInterface
+	currentBlock      ngBlock
+	currentOption     ngOption
+	buf               [24]byte
+	packetBuf         []byte
+	ci                gopacket.CaptureInfo
+	ancil             [1]interface{}
+	blen              int
+	firstSectionFound bool
+	activeSection     bool
+	bigEndian         bool
+}
+
+// NewNgReader initializes a new writer, reads the first section header, and if necessary according to the options the first interface.
+func NewNgReader(r io.Reader, options NgReaderOptions) (*NgReader, error) {
+	ret := &NgReader{
+		r: bufio.NewReader(r),
+		currentOption: ngOption{
+			value: make([]byte, 1024),
+		},
+		options: options,
+	}
+
+	//pcapng _must_ start with a section header
+	if err := ret.readBlock(); err != nil {
+		return nil, err
+	}
+	if ret.currentBlock.typ != ngBlockTypeSectionHeader {
+		return nil, fmt.Errorf("Unknown magic %x", ret.currentBlock.typ)
+	}
+
+	if err := ret.readSectionHeader(); err != nil {
+		return nil, err
+	}
+
+	return ret, nil
+}
+
+// First a couple of helper functions to speed things up
+
+// This is way faster than calling io.ReadFull since io.ReadFull needs an itab lookup, does an additional function call into ReadAtLeast, and ReadAtLeast does additional stuff we don't need
+// Additionally this removes the bounds check compared to io.ReadFull due to the use of uint
+func (r *NgReader) readBytes(buffer []byte) (err error) {
+	var nn int
+	n := uint(0)
+	for n < uint(len(buffer)) {
+		nn, err = r.r.Read(buffer[n:])
+		n += uint(nn)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// The following functions make the binary.* functions inlineable (except for getUint64, which is too big, but not in any hot path anyway)
+// Compared to storing binary.*Endian in a binary.ByteOrder this shaves off about 20% for (ZeroCopy)ReadPacketData, which is caused by the needed itab lookup + indirect go call
+func (r *NgReader) getUint16(buffer []byte) uint16 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint16(buffer)
+	}
+	return binary.LittleEndian.Uint16(buffer)
+}
+
+func (r *NgReader) getUint32(buffer []byte) uint32 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint32(buffer)
+	}
+	return binary.LittleEndian.Uint32(buffer)
+}
+
+func (r *NgReader) getUint64(buffer []byte) uint64 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint64(buffer)
+	}
+	return binary.LittleEndian.Uint64(buffer)
+}
+
+// Now the pcapng implementation
+
+// readBlock reads a the blocktype and length from the file. If the type is a section header, endianess is also read.
+func (r *NgReader) readBlock() (err error) {
```

Comment:
I'm wary of having the returned error be a named return variable named "err".  Lots of Go folks have a ton of muscle memory doing `err := ...` in their code, but if someone does that in a function that returns `(err error)` it has the potential to shadow the return var and cause problems.  I'd lean towards one of the following:

1. just return `error`, and say `err := ...; return err` where you need to
2. use a more descriptive return var name `(returnedErr error)`, to avoid hitting folks' muscle memory

I used to return `(err error)` a ton myself, but I managed to shoot myself in the foot a few times with this, and I've had other folks modifying my code make the same mistakes.

Finally, it's nice to explicitly see a `return nil` at the place in the function where success is returned... otherwise it's non-obvious which is the successful path.  Even seeing a `return` at the end of the function isn't a guarantee... did we reach the end of the function because everything went well, or did we break out of a loop because something went wrong?

Reply:
Normally I don't do that for errors, but I read on the internet that named return values give better code (and there is even a bug open with golang), but I must admit I didn't measure it in this case - and seems unnamed is a tiny bit faster in this case (but with a high p value - so YMMV).
For better clarity I change it.